### PR TITLE
*: add network partition tests

### PR DESCRIPTION
**Pull request by gyuho**
For https://github.com/coreos/etcd/issues/4668.

### CODE REVIEW

**Comment by gyuho** [comment]
@heyitsanthony @xiang90 please review. Thanks!

File: integration/cluster.go [hunk 1]
```diff
@@ -830,3 +830,154 @@ type grpcAPI struct {
 	// Maintenance is the maintenance API for the client's connection.
 	Maintenance pb.MaintenanceClient
 }
+
+// WaitUntilLeaderChange waits until the member at given index gets leader change.
+func (c *ClusterV3) WaitUntilLeaderChange(t *testing.T, idx int) uint64 {
```
Comment (heyitsanthony) [inline integration/cluster.go]: `func (m *member) WaitLeaderChange(t *testing.T) uint64`

File: integration/cluster.go [hunk 2]
```diff
@@ -830,3 +830,154 @@ type grpcAPI struct {
...
+			// TODO: remove connection write timeout by selecting on http response closeNotifier
+			// blocking on https://github.com/golang/go/issues/9524
+			case <-time.After(time.Second + time.Duration(electionTicks)*tickDuration + time.Second + rafthttp.ConnWriteTimeout):
+				t.Fatalf("failed to remove member %s in time", m.s.ID())
+			}
+		}
+	}
+	c.mu.Lock()
+	c.Members = newMembers
+	if err := c.fillClusterForMembers(); err != nil {
```
Comment (heyitsanthony) [inline integration/cluster.go]: why is this necessary? cluster.RemoveMember doesn't do it...

File: integration/cluster.go [hunk 3]
```diff
@@ -830,3 +830,154 @@ type grpcAPI struct {
...
+	}
+	c.clients = newClients
+	c.mu.Unlock()
+
+	csc = toGRPC(c.RandClient()).Cluster
+	mresp, err := csc.MemberList(context.TODO(), &pb.MemberListRequest{}, grpc.FailFast(false))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(mresp.Members) != memberN-1 {
```
Comment (heyitsanthony) [inline integration/cluster.go]: `if len(mresp.Members) != len(newMembers)`?

File: integration/cluster.go [hunk 4]
```diff
@@ -830,3 +830,154 @@ type grpcAPI struct {
...
+	if err = m.Launch(); err != nil {
+		t.Fatal(err)
+	}
+
+	csc = toGRPC(c.RandClient()).Cluster
+	mresp, err := csc.MemberList(context.TODO(), &pb.MemberListRequest{}, grpc.FailFast(false))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(mresp.Members) != memberN+1 {
```
Comment (heyitsanthony) [inline integration/cluster.go]: `if len(mresp.Members) != len(members)`

File: integration/cluster.go [hunk 5]
```diff
@@ -830,3 +830,154 @@ type grpcAPI struct {
...
+		t.Fatalf("len(mresp.Members) expected %d, got %+v", memberN+1, mresp.Members)
+	}
+	return m.s.ID()
+}
+
+// DropConnection drop connections from idx1 (member) to idx2.
+func (c *ClusterV3) DropConnection(t *testing.T, idx1, idx2 int) {
```
Comment (heyitsanthony) [inline integration/cluster.go]: `func (m *member) DropSends(t *testing.T, toId uint64)`

File: integration/cluster.go [hunk 6]
```diff
@@ -830,3 +830,154 @@ type grpcAPI struct {
...
+	m1, m2 := c.Members[idx1], c.Members[idx2]
+	tr1, ok1 := m1.s.Transporter().(*rafthttp.Transport)
+	if !ok1 {
+		t.Fatalf("expected *rafthttp.Transport, got %T", m1.s.Transporter())
+	}
+	tr1.DropSend(m2.s.ID())
+}
+
+// RecoverConnection recovers dropped-connections from idx1 (member) to idx2.
+func (c *ClusterV3) RecoverConnection(t *testing.T, idx1, idx2 int) {
```
Comment (heyitsanthony) [inline integration/cluster.go]: `func (m *member) ResumeSend(t *testing.T, toID uint64)`

File: integration/cluster.go [hunk 7]
```diff
@@ -830,3 +830,154 @@ type grpcAPI struct {
...
+		t.Fatalf("expected *rafthttp.Transport, got %T", m1.s.Transporter())
+	}
+	tr1.RecoverSend(m2.s.ID())
+}
+
+// CreateNetworkPartition drops connections from group src to tgt.
+func (c *ClusterV3) CreateNetworkPartition(t *testing.T, src, tgt []int) {
+	for _, idx1 := range src {
+		for _, idx2 := range tgt {
+			c.DropConnection(t, idx1, idx2)
```
Comment (heyitsanthony) [inline integration/cluster.go]: s/DropConnection/PauseSend/

File: integration/cluster.go [hunk 8]
```diff
@@ -830,3 +830,154 @@ type grpcAPI struct {
...
+	m1, m2 := c.Members[idx1], c.Members[idx2]
+	tr1, ok1 := m1.s.Transporter().(*rafthttp.Transport)
+	if !ok1 {
+		t.Fatalf("expected *rafthttp.Transport, got %T", m1.s.Transporter())
+	}
+	tr1.RecoverSend(m2.s.ID())
+}
+
+// CreateNetworkPartition drops connections from group src to tgt.
+func (c *ClusterV3) CreateNetworkPartition(t *testing.T, src, tgt []int) {
```
Comment (heyitsanthony) [inline integration/cluster.go]: `func (m *member) PausePartition(t *testing.T, tgts []uint64)`

File: integration/cluster.go [hunk 9]
```diff
@@ -830,3 +830,154 @@ type grpcAPI struct {
...
+	for _, idx1 := range src {
+		for _, idx2 := range tgt {
+			c.DropConnection(t, idx1, idx2)
+		}
+	}
+}
+
+// RecoverNetworkPartition recovers dropped-connections from group src to tgt.
+func (c *ClusterV3) RecoverNetworkPartition(t *testing.T, src, tgt []int) {
```
Comment (heyitsanthony) [inline integration/cluster.go]: `func (m *member) ResumePartition(t *testing.T, tgts []uint64)`

File: integration/cluster.go [hunk 10]
```diff
@@ -830,3 +830,154 @@ type grpcAPI struct {
...
+	case <-donec:
+	case <-time.After(timeout):
+		close(stopc)
+	}
+
+	return m.s.Lead()
+}
+
+// RemoveMember removes and terminates the member with given id from cluster.
+func (c *ClusterV3) RemoveMember(t *testing.T, id uint64) {
```
Comment (xiang90) [inline integration/cluster.go]: do we really need these? can we use the existing memberadd, member remove?

File: etcdserver/server.go [hunk 11]
```diff
@@ -511,6 +511,8 @@ func (s *EtcdServer) Cluster() *membership.RaftCluster { return s.cluster }
 
 func (s *EtcdServer) RaftHandler() http.Handler { return s.r.transport.Handler() }
 
+func (s *EtcdServer) Transporter() rafthttp.Transporter { return s.r.transport }
```
Comment (xiang90) [inline etcdserver/server.go]: can we do something similar to what we did for Pause and Resume?

**Comment by xiang90** [comment]
@gyuho There are too much stuff in one PR.

Can we do this step by step?

I think the first step is to add isolation functionality to rafthttp, so that a peer can only drop message send/recv from a given peer.

File: integration/network_partition_test.go [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+func TestNetworkPartition(t *testing.T) {
+	defer testutil.AfterTest(t)
+
+	clus := NewClusterV3(t, &ClusterConfig{Size: 5})
+	defer clus.Terminate(t)
+
+	leadIdx1 := clus.WaitLeader(t)
+
+	// group A: leader, follower / group B: follower, follower, follower
+	groupA := []int{leadIdx1, (leadIdx1 + 1) % 5}
```
Comment (heyitsanthony) [inline integration/network_partition_test.go]: s/groupA/minorityPartition?

File: integration/network_partition_test.go [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+	groupB := []int{(leadIdx1 + 2) % 5, (leadIdx1 + 3) % 5, (leadIdx1 + 4) % 5}
```
Comment (heyitsanthony) [inline integration/network_partition_test.go]: s/groupB/majorityPartition?

File: integration/network_partition_test.go [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+	"testing"
+
+	"golang.org/x/net/context"
+
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/pkg/testutil"
+	"github.com/coreos/etcd/pkg/types"
+)
+
+func TestNetworkPartition(t *testing.T) {
```
Comment (heyitsanthony) [inline integration/network_partition_test.go]: split these into separate tests?

**Comment by gyuho** [comment]
@xiang90 @heyitsanthony Ok I will separate the PR and address the suggestions. Thanks!

File: rafthttp/transport.go [hunk 15]
```diff
@@ -170,16 +170,20 @@ func (t *Transport) Send(msgs []raftpb.Message) {
 		t.mu.RUnlock()
 
 		if pok {
-			if m.Type == raftpb.MsgApp {
-				t.ServerStats.SendAppendReq(m.Size())
+			if !p.isPaused() {
```
Comment (heyitsanthony) [inline rafthttp/transport.go]: isn't this already handled in `peer.send()`?

File: rafthttp/transport.go [hunk 16]
```diff
@@ -170,16 +170,20 @@ func (t *Transport) Send(msgs []raftpb.Message) {
...
+				continue
 			}
-			p.send(m)
-			continue
 		}
 
 		if rok {
-			g.send(m)
-			continue
+			if !g.paused {
```
Comment (heyitsanthony) [inline rafthttp/transport.go]: check `paused` in `send`?

File: rafthttp/transport.go [hunk 17]
```diff
@@ -206,6 +210,34 @@ func (t *Transport) Stop() {
 	t.remotes = nil
 }
 
+// DropSend drops messages to the specified peer.
+func (t *Transport) DropSend(id types.ID) {
```
Comment (heyitsanthony) [inline rafthttp/transport.go]: `PausePeer`?

File: rafthttp/transport.go [hunk 18]
```diff
@@ -206,6 +210,34 @@ func (t *Transport) Stop() {
...
+	}
+	g, gok := t.remotes[id]
+	if gok {
+		g.paused = true
+	}
+	t.mu.RUnlock()
+}
+
+// RecoverSend recovers the send pause to the given peer.
+func (t *Transport) RecoverSend(id types.ID) {
```
Comment (heyitsanthony) [inline rafthttp/transport.go]: `ResumePeer`?

File: rafthttp/transport.go [hunk 19]
```diff
@@ -206,6 +210,34 @@ func (t *Transport) Stop() {
 	t.remotes = nil
 }
 
+// DropSend drops messages to the specified peer.
+func (t *Transport) DropSend(id types.ID) {
+	t.mu.RLock()
+	p, pok := t.peers[id]
+	if pok {
+		p.setPaused(true)
```
Comment (heyitsanthony) [inline rafthttp/transport.go]: p.(Pausable).Pause()

File: rafthttp/transport.go [hunk 20]
```diff
@@ -206,6 +210,34 @@ func (t *Transport) Stop() {
...
+	}
+	t.mu.RUnlock()
+}
+
+// RecoverSend recovers the send pause to the given peer.
+func (t *Transport) RecoverSend(id types.ID) {
+	t.mu.RLock()
+	p, pok := t.peers[id]
+	if pok {
+		p.setPaused(false)
```
Comment (heyitsanthony) [inline rafthttp/transport.go]: p.(Pausable).Resume()

File: rafthttp/transport.go [hunk 21]
```diff
@@ -206,6 +210,34 @@ func (t *Transport) Stop() {
...
+// DropSend drops messages to the specified peer.
+func (t *Transport) DropSend(id types.ID) {
+	t.mu.RLock()
+	p, pok := t.peers[id]
+	if pok {
+		p.setPaused(true)
+	}
+	g, gok := t.remotes[id]
+	if gok {
+		g.paused = true
```
Comment (heyitsanthony) [inline rafthttp/transport.go]: have `remote` implement the `Pausable` interface instead?

**Comment by heyitsanthony** [comment]
approach looks OK, but interfaces/naming need some polishing

File: rafthttp/transport.go (hunk 21 above)
Comment (gyuho) [inline rafthttp/transport.go]: Not now. I will try adding `Resume` method to `remote` to satisfy `Pausable` interface here https://github.com/coreos/etcd/blob/master/rafthttp/remote.go#L59-L61.

Thanks!

File: integration/cluster.go (hunk 10 above)
Comment (gyuho) [inline integration/cluster.go]: We can use current member add and remove method to cluster, but it will use v2 API.

Comment (xiang90) [inline integration/cluster.go]: that is fine. we can refactor it later.

Comment (gyuho) [inline integration/cluster.go]: Ok sounds good. Will do that.

**Comment by gyuho** [comment]
@xian90 @heyitsanthomy PTLA. Thanks!

File: etcdserver/server.go [hunk 22]
```diff
@@ -775,6 +775,22 @@ func (s *EtcdServer) triggerSnapshot(ep *etcdProgress) {
 	ep.snapi = ep.appliedi
 }
 
+// CutPeer drops messages to the specified peer.
```
Comment (xiang90) [inline etcdserver/server.go]: Can we do something similar to 

```go
// Pausable is a testing interface for pausing transport traffic.
type Pausable interface {
	Pause()
	Resume()
}
```
?

File: integration/cluster.go [hunk 23]
```diff
@@ -734,6 +734,52 @@ func (m *member) Metric(metricName string) (string, error) {
...
+	m.s.CutPeer(toMember.s.ID())
+}
+
+func (m *member) ResumeSend(t *testing.T, toMember *member) {
+	m.s.MendPeer(toMember.s.ID())
+}
+
+// PausePartition drops connections to tgts.
+func (m *member) PausePartition(t *testing.T, tgts []*member) {
+	for _, tgt := range tgts {
```
Comment (xiang90) [inline integration/cluster.go]: this should be bi-direction. only pause one direction is not enough. right?

File: integration/network_partition_test.go [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+	"time"
+
+	"golang.org/x/net/context"
+
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/pkg/testutil"
+	"github.com/coreos/etcd/pkg/types"
+)
+
+func TestNetworkPartition5NodeLeaderInMinority(t *testing.T) {
```
Comment (xiang90) [inline integration/network_partition_test.go]: 5Members

File: integration/network_partition_test.go [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+	defer testutil.AfterTest(t)
+
+	clus := NewClusterV3(t, &ClusterConfig{Size: 5})
+	defer clus.Terminate(t)
+
+	leadIdx1 := clus.WaitLeader(t)
```
Comment (xiang90) [inline integration/network_partition_test.go]: just leadIndex?

File: integration/network_partition_test.go [hunk 26]
```diff
@@ -0,0 +1,152 @@
...
+
+	// minority: leader, follower / majority: follower, follower, follower
+	minority := []int{leadIdx1, (leadIdx1 + 1) % 5}
+	majority := []int{(leadIdx1 + 2) % 5, (leadIdx1 + 3) % 5, (leadIdx1 + 4) % 5}
+	tgts := make([]*member, len(majority))
```
Comment (xiang90) [inline integration/network_partition_test.go]: what is tgt?

File: integration/network_partition_test.go [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+	for i, idx := range majority {
+		tgts[i] = clus.Members[idx]
+	}
+
+	// cut connections from minority to majority
+	for _, idx := range minority {
+		clus.Members[idx].PausePartition(t, tgts)
+	}
+
```
Comment (xiang90) [inline integration/network_partition_test.go]: should we cut majority to minority too?

File: integration/network_partition_test.go [hunk 28]
```diff
@@ -0,0 +1,152 @@
...
+	// minority with leader can't reach quorum, so revert back follower
+	clus.Members[leadIdx1].WaitLeaderChange(t)
+
+	// new leader must be from majority
+	donec := make(chan struct{})
+	tm := clus.Members[majority[0]].s.Cfg.ReqTimeout()
+	go func() {
+		defer close(donec)
+		clus.waitLeader(t, tgts)
```
Comment (xiang90) [inline integration/network_partition_test.go]: just do waitLeader is good enough i guess...

File: integration/network_partition_test.go [hunk 29]
```diff
@@ -0,0 +1,152 @@
...
+	}()
+	select {
+	case <-donec:
+	case <-time.After(tm):
+		t.Fatalf("leader election took too long (%v)", tm)
+	}
+
+	// write to majority to make it more up-to-date
+	kvc := toGRPC(clus.Client(majority[0])).KV
```
Comment (xiang90) [inline integration/network_partition_test.go]: i think we have a ensure progress method or something like that already.

File: integration/network_partition_test.go [hunk 30]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	// recover network partition
+	for _, idx := range minority {
+		clus.Members[idx].ResumePartition(t, tgts)
+	}
+
+	// ensure leader in majority still holds
+	leadIdx2 := clus.WaitLeader(t)
+	inMajority := false
```
Comment (xiang90) [inline integration/network_partition_test.go]: this is not necessary. waitleader should already ensure everyone agrees on the same leader, no?

File: integration/network_partition_test.go (hunk 25 above)
Comment (gyuho) [inline integration/network_partition_test.go]: ok will fix. thanks.

File: etcdserver/server.go (hunk 22 above)
Comment (gyuho) [inline etcdserver/server.go]: @xiang90 Do you mean `Pause(types.ID)`? We already have `PauseSending` and `ResumeSending` methods for `EtcdServer`.

```go
func (r *raftNode) pauseSending() {
	p := r.transport.(rafthttp.Pausable)
	p.Pause()
}

func (r *raftNode) resumeSending() {
	p := r.transport.(rafthttp.Pausable)
	p.Resume()
}
```

But this doesn't pause the remotes.

```go
func (t *Transport) Pause() {
	for _, p := range t.peers {
		p.(Pausable).Pause()
	}
}

func (t *Transport) Resume() {
	for _, p := range t.peers {
		p.(Pausable).Resume()
	}
}
```

File: integration/network_partition_test.go (hunk 26 above)
Comment (gyuho) [inline integration/network_partition_test.go]: `tgt` means `target`

**Comment by gyuho** [comment]
@xiang90 @heyitsanthony All addressed. PTAL. Thanks.

File: integration/cluster.go [hunk 31]
```diff
@@ -734,6 +734,52 @@ func (m *member) Metric(metricName string) (string, error) {
...
+	select {
+	case <-donec:
+	case <-time.After(timeout):
+		close(stopc)
+	}
+
+	return m.s.Lead()
+}
+
+func (m *member) PauseSend(t *testing.T, toMember *member) {
```
Comment (heyitsanthony) [inline integration/cluster.go]: are these functions necessary? just call m.s.CutPeer/MendPeer directly? The send part is sort of misleading because it's also rejecting messages send from the peer too.

File: integration/cluster.go (hunk 23 above)
Comment (heyitsanthony) [inline integration/cluster.go]: it looks bi-directional, the `PauseSend` is `CutPeer` and that cuts sends and receives, the naming is just confusing...

File: integration/network_partition_test.go [hunk 32]
```diff
@@ -0,0 +1,156 @@
...
+	clus := NewClusterV3(t, &ClusterConfig{Size: 5})
+	defer clus.Terminate(t)
+
+	leadIndex := clus.WaitLeader(t)
+
+	// minority: leader, follower / majority: follower, follower, follower
+	minority := []int{leadIndex, (leadIndex + 1) % 5}
+	majority := []int{(leadIndex + 2) % 5, (leadIndex + 3) % 5, (leadIndex + 4) % 5}
+
+	minorityMembers := make([]*member, len(minority))
```
Comment (heyitsanthony) [inline integration/network_partition_test.go]: have a helper function `func partitionByIdx(c *cluster, idxs []int) []*member`?

File: integration/cluster.go (hunk 31 above)
Comment (gyuho) [inline integration/cluster.go]: Yeah now seems confusing to me as well. Will remove.

File: integration/network_partition_test.go [hunk 33]
```diff
@@ -0,0 +1,156 @@
...
+	clus := NewClusterV3(t, &ClusterConfig{Size: 4})
+	defer clus.Terminate(t)
+
+	leadIndex := clus.WaitLeader(t)
+
+	// groupA: leader, follower / groupB: follower, follower
+	groupA := []int{leadIndex, (leadIndex + 1) % 4}
+	groupB := []int{(leadIndex + 2) % 4, (leadIndex + 3) % 4}
+
+	groupAMembers := make([]*member, len(groupA))
```
Comment (heyitsanthony) [inline integration/network_partition_test.go]: s/groupAMembers/leaderPartition
s/groupBMembers/followerPartition
?

File: integration/network_partition_test.go [hunk 34]
```diff
@@ -0,0 +1,156 @@
...
+	}
+
+	// minority with leader can't reach quorum, so revert back follower
+	clus.Members[leadIndex].WaitLeaderChange(t)
+
+	// new leader must be from majority
+	clus.waitLeader(t, majorityMembers)
+
+	// recover network partition
+	for _, idx := range minority {
```
Comment (heyitsanthony) [inline integration/network_partition_test.go]: helper function `func resumePartition(t *testing.T, clus *cluster, part []*member)`?

**Comment by heyitsanthony** [comment]
lgtm aside from minor nits; defer to @xiang90

File: integration/network_partition_test.go [hunk 35]
```diff
@@ -0,0 +1,134 @@
...
+
+	// minority: leader, follower / majority: follower, follower, follower
+	minority := []int{leadIndex, (leadIndex + 1) % 5}
+	majority := []int{(leadIndex + 2) % 5, (leadIndex + 3) % 5, (leadIndex + 4) % 5}
+
+	minorityMembers := partitionByIdx(clus.cluster, minority)
+	majorityMembers := partitionByIdx(clus.cluster, majority)
+
+	// network partition (bi-directional)
+	pausePartition(t, minorityMembers, majorityMembers)
```
Comment (xiang90) [inline integration/network_partition_test.go]: can we just call this partition(A, B). Also partition(A, B) should partition both A and B. We should not need to call P(A, B) and P(B, A)

File: integration/network_partition_test.go [hunk 36]
```diff
@@ -0,0 +1,134 @@
...
+		t.Fatalf("%d expected no leader, got %s", leadIndex, types.ID(id))
+	}
+
+	// recover network partition
+	resumePartition(t, leaderPartition, followerPartition)
+	resumePartition(t, followerPartition, leaderPartition)
+	clusterMustProgress(t, clus.Members)
+}
+
+func partitionByIdx(clus *cluster, idxs []int) []*member {
```
Comment (xiang90) [inline integration/network_partition_test.go]: getMembersByIndexSlice

File: integration/cluster.go (hunk 23 above)
Comment (xiang90) [inline integration/cluster.go]: No. I mean this func should like

```go

func (m *member) PartitionWith(t *testing.T, others []*member) {
    for _, other := range others {
        other.s.CutPeer(m.s.ID())
        m.s.CutPeer(other.s.ID())
    }
}
```

**Comment by gyuho** [comment]
@heyitsanthony @xiang90 Addressed. PTAL. Thanks.

File: integration/network_partition_test.go [hunk 37]
```diff
@@ -0,0 +1,128 @@
...
+
+	// minority: leader, follower / majority: follower, follower, follower
+	minority := []int{leadIndex, (leadIndex + 1) % 5}
+	majority := []int{(leadIndex + 2) % 5, (leadIndex + 3) % 5, (leadIndex + 4) % 5}
+
+	minorityMembers := getMembersByIndexSlice(clus.cluster, minority)
+	majorityMembers := getMembersByIndexSlice(clus.cluster, majority)
+
+	// network partition (bi-directional)
+	injectPartitionWith(t, minorityMembers, majorityMembers)
```
Comment (xiang90) [inline integration/network_partition_test.go]: probably just injectPartition?

File: integration/network_partition_test.go [hunk 38]
```diff
@@ -0,0 +1,128 @@
...
+	minority := []int{(leadIndex + 3) % 5, (leadIndex + 4) % 5}
+
+	minorityMembers := getMembersByIndexSlice(clus.cluster, minority)
+	majorityMembers := getMembersByIndexSlice(clus.cluster, majority)
+
+	// network partition (bi-directional)
+	injectPartitionWith(t, minorityMembers, majorityMembers)
+
+	// leader in majority must hold
+	leadIndex2 := clus.WaitLeader(t)
```
Comment (xiang90) [inline integration/network_partition_test.go]: we need to wait for two election timeouts here. the minority should lose its leader.

if we do not wait, nothing will change since the timeout is not exceeded.

**Comment by gyuho** [comment]
@xiang90 Just fixed and added sleeps with election timeouts when checking leadership.

PTAL.

Thanks.

File: integration/network_partition_test.go [hunk 39]
```diff
@@ -0,0 +1,138 @@
...
+	minorityMembers := getMembersByIndexSlice(clus.cluster, minority)
+
+	// network partition (bi-directional)
+	injectPartition(t, majorityMembers, minorityMembers)
+
+	electionTimeout := time.Duration(majorityMembers[0].s.Cfg.ElectionTicks) * time.Millisecond
+	time.Sleep(2 * electionTimeout)
+
+	// leader in majority must hold
+	leadIndex2 := clus.WaitLeader(t)
```
Comment (xiang90) [inline integration/network_partition_test.go]: is this correct? the minority should lost leader, right? clus is the whole cluster.

**Comment by gyuho** [comment]
@xiang90 You are right. Just fixed it to check waitLeader within majority for that case. PTAL.

Thanks!

**Comment by gyuho** [comment]
As Xiang suggested, will make the leader checking more strict.

**Comment by gyuho** [comment]
@xiang90 Cleaned up as you suggested. PTAL. Thanks!

File: integration/network_partition_test.go [hunk 40]
```diff
@@ -0,0 +1,140 @@
...
+	injectPartition(t, leaderPartition, followerPartition)
+
+	// no group has quorum, so leader must be lost in all members
+	clus.WaitNoLeader(t)
+
+	// recover network partition (bi-directional)
+	recoverPartition(t, leaderPartition, followerPartition)
+
+	// wait extra election timeout
+	time.Sleep(2 * leaderPartition[0].electionTimeout())
```
Comment (xiang90) [inline integration/network_partition_test.go]: remove this? mustprogress should do the wait internally, no?

**Comment by xiang90** [comment]
lgtm

File: integration/network_partition_test.go (hunk 40 above)
Comment (gyuho) [inline integration/network_partition_test.go]: just remove. thanks

**Comment by gyuho** [comment]
Merging after greens.
